sycl/include/CL/sycl/detail/spirv_glue.hpp

```diff
@@ -0,0 +1,60 @@
+//==-------------- spirv_glue.hpp - SYCL kernel side init via SPIRV --------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#ifdef __SYCL_DEVICE_ONLY__
+
+#include "CL/__spirv/spirv_vars.hpp"
+
+#define DEFINE_INIT_SIZES(POSTFIX)                                             \
```

Comment:
This may be a silly request, but would it be possible to call this something other than spirv_glue.hpp, as at the moment we currently use these functions as a sort of top level API that offloads to SPIRV/SPIR intrinsics based on what we compile for. e.g: https://github.com/triSYCL/sycl/blob/sycl/unified/next/sycl/include/CL/__spir/spir_vars.hpp I'm aware the names are SPIRV oriented though.

Maybe this isn't the place for that though or the best implementation on our end (I did have a slightly more complex/generic implementation in a prior commit but it was a bit overly complex for our current needs), but it would be nice to have a sort of top level id/size/group etc. API that calls the underlying implementation be it OpenCL SPIR calls, SPIRV calls or something else entirely. Hopefully this is a somewhat reasonable sounding idea/request.

Reply:
Good point. We should not be tied to SPIRV, as there can be other models or even runtimes on the device side underneath. Is it correct interpretation of your concern? How about device_rt_intrin.hpp or device_rt_builtins.hpp? For now there is __spirv* variant only, later implementation for more native RTs can be added. @bader, please also comment.

Reply:
That is indeed a correct interpretation of my concern! device_rt_builtins.hpp sounds fine by me but if you want to call it device_rt_intrin.hpp that's also all good in my book. Thanks for taking my concern into consideration!

Reply:
I'm totally agree with @agozillon that `spirv_glue` is unfortunate name - even w/o applying his "generalization idea".
I think current implementation of these functions can reside in `spirv_vars` header.

I'm all in for using built-ins to get id/size information. 
@agozillon, feel free to propose any changes simplifying triSYCL development. I think we can re-use https://github.com/triSYCL/sycl/blob/sycl/unified/next/sycl/include/CL/__spir/spir_vars.hpp and have two different implementations of `__spir_ocl_get_*` built-ins:
```c++
#if __COMPILE_TO_SPIRV__
typedef size_t size_t_vec __attribute__((ext_vector_type(3)));
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalSize;
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalInvocationId;
extern "C" const __constant size_t_vec __spirv_BuiltInWorkgroupSize;
extern "C" const __constant size_t_vec __spirv_BuiltInLocalInvocationId;
extern "C" const __constant size_t_vec __spirv_BuiltInWorkgroupId;
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalOffset;

// __spir_ocl_get_* implementations can be generated by macro function
constexpr __spir_ocl_get_global_id(int dim) {
  switch (dim) {
  case 0: 
    return __spirv_BuiltInGlobalInvocationId.x;
  case 1: 
    return __spirv_BuiltInGlobalInvocationId.y;
  case 2: 
    return __spirv_BuiltInGlobalInvocationId.z;
  }
}
...
#elif __XILINX_SPECIFIC_IMPLEMENTATION__
  // some other implementation of __spir_ocl_get_*
#endif
```

Reply:
@bader, I already made the change to device_rt_intrin.hpp, which seems to address @agozillon's concerns. Are you OK if we live with this name for now?

Reply:
This PR became a little bit messy. Now first commit adds spirv_glue.hpp file, which is later renamed to device_rt_intrin.hpp in the third commit of the pull request. 
Now I have a few options how to merge this pull request, but none of them look good to me. I can squash all the commits into single commit, but in this case first two commits (not directly related to the feature) will be part of the hierarchical parallelism commit. Another option is to save current history with merge or rebase, but it also looks strange because first patch comment about renaming is addressed in the third commit.

I really like the separation of first two commits because it's makes review easier and keeps history clean.
@kbobrovs, could you send first two commits not related to hierarchical parallelism in a separate pull request, please?
NOTE: it means that file with `init*` should be named device_rt_intrin.hpp in the first commit.

Reply:
@bader, OK, will do

Reply:
Thank you!

Reply:
Something along those lines would be useful I think, I don't necessarily mind what form it takes and the top level function doesn't need to be named __spir_ocl_get_*  it could be __sycl_get_* if you wish, in our case that's just my cheap/cheat way of working around conflicts with user functions with the same mangling as OpenCL SPIR built-ins like get_global_id etc.

If you guys are happy with an abstraction like that (similar to our current approach) it's good with us, but perhaps someone has a more generic approach? As for the most part it would be nice to have an API for these functions that allows anyone to add there own underlying implementation, not necessarily just SPIR/SPIRV built-ins!

I had a slightly more generic and perhaps less sane implementation in a previous commit: https://github.com/triSYCL/sycl/commit/489f98d16adb07a3b4a805865b348b0ed8a6d52b#diff-0e3fe0cefdf466dc994182d2e123ba01 that may be more useful? I reverted it as we don't really have need for other built-ins just yet, so I can't really justify forcing someone else to maintain it... and a separate file approach allowed me to avoid frequent merge conflicts.

In either case, I don't want to be a bother and hold up these commits for too long, it can perhaps be a separate commit/merge/issue discussion. For now I'm happy with either, the renaming of the header to something less SPIRV centric or the movement of the functions to reside in spirv_vars. Whatever works for you guys!

Reply:
Introducing _sycl_get* (or __sycl_get*) C APIs (C  mangling assumed, right?) sounds like a good idea. Those would be generic SYCL "intrinsics" which must be resolved by actual  SYCL device Back-end, e.g. by linking with <be>_device.bc which will provide implementations of __sycl_get* for <be>. So far we have __spirv* serving in this role, with a nice side effect that we don't need to provide <spirv>_device.bc, because SPIRV translator knows how to lower them. But the problem with current __spirv* is that the interface is based on variables rather than functions, which is less robust.

This sounds like a separate important task which needs little more discussion, so I suggest that for the purpose of hierarchical parallelism I follow what @bader  suggested - move init functions in the spirv_vars.hpp for now removing the device_rt_intrin.hpp, then we can refactor along the lines @agozillon proposed.
?

Reply:
Yes, that sounds like a separate task. I am happy with the change of removing the new header and moving the functions into spirv_vars.hpp for now, that seems fine.